### ISSUE TITLE
Latency spike for high-frequency oscillator with other settings

### ISSUE THREAD

**Issue by j5v** [surge-synthesizer/surge#2387]
**Describe the bug**
Dropout (maybe Surge has a latency spike; everything in the host stops, but it's recoverable) when 
Previously, a similar scenario would cause the host to crash.
This is an initial report; I hope to revisit to provide minimal reproducible steps.

**Please let us know your surge version**
Version: 7.0.nightly.8089a93
(sorry)
VST3
64-bit
Windows

**To Reproduce**
1. Start with init patch
2. Set:
    - oscillator type: S+H,
    - **Sync** high (44)
    - **Scene** octave +3,
    - **Osc** octave is +3,
    - **Detune** max,
    - **Unison voices** max.
3. Play a high note in the host, say C7.
4. The host experiences a dropout, and stops all sound.

**Expected behavior**
White-ish noise should play.

**Screenshots**
![S+H boundary](https://user-images.githubusercontent.com/13245475/88839215-aa605a00-d1d2-11ea-8d7c-61b9df7f4a8e.PNG)
Boundary shown

![dropout](https://user-images.githubusercontent.com/13245475/88839323-d24fbd80-d1d2-11ea-97cc-dc57753f7776.PNG)
Cakewalk's dropout message.

**Desktop (please complete the following information):**
Cakewalk, latest as at mid-July,
Windows 10.

**Additional context**
I was deliberately driving the S+H sampling high, but think I found an edge case for the wavetable cursor.
Changing values of **Sync** reveal a boundary. On my system, 44 is OK; 45 causes gaps in the audio; 46 causes Cakewalk to stop all. Your mileage may vary.

**Comment by VincyZed** [comment]
Just tested this on FL Studio, Windows 10. I get very high CPU usage (like 75%) with Sync set at 44.
However, I get FL Studio to max out and sometimes crash too when I set Sync to maximum.

**Comment by baconpaul** [comment]
I wonder if my recent sync fix causes this. Can you let me know if it happens with 1.7.0 also?

**Comment by VincyZed** [comment]
Yes, still happens in 1.7.0.

**Comment by baconpaul** [comment]
cool so 1.7.0 and latest nightly both crash this way. thanks.

**Comment by j5v** [comment]
Speculation: (without looking at the code, which paul is more familiar with!) I suspect the oscillator sweeps across the wave with a maximum increment (maybe 1.0?), so with a high frequency, many samples must be made, even though they're not used directly.

The current situation is much less critical than the all-out 'hang' that happened before @baconpaul 's intervention.

**Comment by baconpaul** [comment]
Yeah that's basically right although that's not how the algorithm works. At a very high pitch the number of blits per second goes up so we have to convolve the blit stream many more times to fill the buffer. 

Normally we need to convolve once or twice per block. With S&H with sync at 0, octave at +3, and pitch at +48 (extended) you have about 26 convolutions.

If I change the sync in that situation to 60 I get 845 convolutions.

And if I set the unison all the way up with the unison detune all the way up, I get 16 * 1000

that will be ... CPU intensive. My mac doesn't drop out with one note. But with 5 I start getting some drops and my fan comes on. But I have a brand new MacBook pro etc...

Some choices

1: Limit the top of the high frequency (or floor the time between blits)
2: Apply the "doctor joke" and update the manual. (https://www.goodreads.com/quotes/7570350-the-patient-says-doctor-it-hurts-when-i-do-this)
3: Do some third thing I can't think of right now but I'm all ears!

**Comment by j5v** [comment]
Given that it's dependent on the performance of the host computer, and that this is an edge case, I'd be happy with the **doctor principle**.

I wouldn't want to exclude some scenarios, with a limit that doesn't precisely correlate with the cause, and is set to the lowest common denominator to ensure consistently for everyone (it won't be that effective). It's like a host limiting overall polyphony because someone might be running it on a Pi.

**Comment by mkruselj** [comment]
I think it makes no sense to calculate convolute if the frequency is outside of [0 Hz, Nyquist] range. And even 0 Hz should not be allowed since it's not good for speakers.

**Comment by baconpaul** [comment]
Yeah that's the right cap. I think I can calculate that reliably in "t" space.

**Comment by mkruselj** [comment]
Maybe 10 Hz should be lowest cap?

**Comment by baconpaul** [comment]
oh the long time one is less a worry. it's the short time one i need to calculate the limit for.

**Comment by baconpaul** [comment]
(short time == high frequency)

**Comment by mkruselj** [comment]
Yeah sure, I was just thinking we shouldn't generate close to DC signals. Highpass filter is at lowest 6 Hz, but its rolloff doesn't sort out that low end junk really, so I was thinking oscillators shouldn't even product stuff below 10 or so Hz.

**Comment by mkruselj** [comment]
Even shorter version - let's doctor joke it @baconpaul?

**Comment by baconpaul** [comment]
Not yet.

**Comment by mkruselj** [comment]
@baconpaul Claude is finding some _really_ interesting stuff for this one, too. (Still percolating...)

**Comment by mkruselj** [comment]
I dug into this again and measured it properly. Short version: the S&H CPU blowup is a small, local fix, but where exactly to put the clamp is a sound-vs-CPU trade-off that bundled patches force on us. While at it I also checked Classic (fine) and Wavetable (has a similar problem, with an easy fix).

All numbers below are offline renders through surgepy at 44.1 kHz, reported as CPU seconds per second of audio (1.0 = one core fully busy), from a clang Release build on Windows. Absolute figures will vary by machine; the ratios are what matter.

## S&H: root cause

In `convolute()`, the time to the next step is `t = note_to_pitch_inv(detune + l_sync.v)` ([SampleAndHoldOscillator.cpp#L247](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/oscillators/SampleAndHoldOscillator.cpp#L247)), and each `convolute()` call advances the voice by about `t/2`. Nothing bounds `t`:

- `process_block` clamps `pitch` to 148, but sync and unison detune are added on top of that inside `convolute`, so the effective rate is uncapped. This is the unison problem mentioned above.
- `l_sync` is only floored at 0, never capped, so modulating sync past 60 keeps doubling the work every 12 semitones.
- Classic doesn't blow up because it has `sync = min(l_sync.v, (12 + 72 + 72) - pitch)` ([ClassicOscillator.cpp#L392](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/oscillators/ClassicOscillator.cpp#L392)), which dates back to the original code drop. S&H never got an equivalent.

Once a voice produces more than one step per oversampled sample, the extra steps contribute nothing except lowering the level. Band-limited S&H noise at event rate R keeps only about fs/R of its power in band, so it loses 3 dB per octave. The measurements match that exactly (report settings: key 96, osc and scene octave +3, detune at max):

| unison, sync | CPU | level |
|---|---|---|
| 16 voices, sync 0 | 0.041 | −21.8 dB |
| 16 voices, sync 36 | 0.282 | −30.5 dB |
| 16 voices, sync 60 | **1.083** | −36.6 dB |
| 5 notes × 16 voices, sync 60 | **5.73** | −28.0 dB |

CPU doubles and level drops 3 dB for every octave of sync.

## S&H: proposed approach

Clamp `t` **per voice, after detune and sync are applied**, so a cycle can't get shorter than K oversampled samples. Then apply the level drop the skipped steps would have caused (`sqrt(t / tmin)`), instead of computing them:

```cpp
float levelComp = 1.f;
const float tmin = K * pitchmult; // shortest allowed cycle (two steps), in oversampled samples

if (t < tmin)
{
    levelComp = sqrt(t / tmin);
    t = tmin;
}
...
float randt = rand11 * (1 - wfabs) - wf * last_level2[voice]; // correlation keeps using unscaled levels
randt = randt * mech::rcp(1.0f - wfabs);
randt = min(0.5f, max(-0.5f, randt));
last_level2[voice] = randt;
randt *= levelComp;
```

This is different from clamping pitch, which is why I think it avoids the "a simple clamp isn't the answer" problem:

- **Unison:** it's per voice and includes detune.
- **Sample rate:** it adapts automatically, because `pitchmult` already accounts for it.
- **Modulation:** it also bounds sync pushed past 60.
- **Unclamped settings:** when it doesn't kick in, `levelComp` is 1 and `last_level2 == last_level`, so the output is identical to today. (`last_level2` already exists and is zeroed in `init()`, but nothing reads it at the moment.)

With K=2 (cycle frequency capped at the oversampled Nyquist):

| case | CPU before → after | level before → after |
|---|---|---|
| 16 voices, sync 60 | 1.083 → 0.057 | −36.6 → −37.0 |
| 5 notes × 16 voices, sync 60 | 5.73 → 0.26 | −28.0 → −29.9 |
| 16 voices, sync 60, 96 kHz | 1.117 → 0.112 | −33.2 → −33.5 |
| key 60, 16 voices, sync 60 | 0.434 → 0.052 | −32.6 → −32.8 |
| correlation 0 and ±50%, sync 24–60 | ~0.066 → 0.009 | within 0.3–2 dB |

(The 5-note run looks like a level change, but it isn't: the unclamped build's polyphonic output varies by several dB between identical runs, see side finding 1 below.)

**±100% correlation doesn't match.** There the output isn't noise: the ±0.5 clip holds the level at a fixed value, so you get a square wave, and noise-based compensation gets it wrong. At −100%, today's output is a constant −18.4 dB tone at any sync above 0 (and silent at sync 0); with K=2 it becomes −24 dB at sync 24 and −33 dB at sync 60. I tried scaling only the random part instead. That keeps the square wave, but plain noise comes out 3 dB too loud, because scaling before the clip changes how much gets clipped.

## S&H: bundled patches rely on this range

This is not a corner case. Of the 3,561 bundled patches, 260 S&H oscillators exist and 168 have sync above 0. At 44.1k the K=2 clamp would already kick in at key ≤ 72 for 20 of them and at key ≤ 108 for 73 of them, before counting pitch envelopes. Most are 3rd-party drums (Slowboat and Cybersoda snares/kicks, John Valentine claps, Dan Maurer rhythms), and 27 of those 73 have |correlation| ≥ 50%. Only two factory patches are affected, both at very high keys (Spooky Fish, Organ 1).

Rendering the affected patches with the clamp off and at several values of K (lower K kicks in higher up), the result is clear. The compensation holds for near-zero correlation, but correlated S&H shifts. Above about ±62% correlation the correlation recursion is effectively unstable and the clip is what shapes the output, so no closed-form correction will match it. Snare 1, Saturn V and Hats 1 were rendered 8 times each, with run-to-run std of 0.0–0.7 dB in the bands listed:

| patch (correlation) | K=2 | K=0.5 | K=0.125 |
|---|---|---|---|
| Slowboat Snare 1 at MIDI 72 (one osc −0.66) | 8–20 kHz −2.7 to −4.5 dB | ≤ 0.8 dB | ≤ 0.5 dB |
| Inigo Kennedy Saturn V at MIDI 96 (−0.9) | 500–2k −2.0, 8–12k −2.4 | ≤ 0.5 | ≤ 1.0 |
| Slowboat Hats 1 at MIDI 108 (−1.0) | −2.2 in 500–2k | ≤ 1.3 in weak bands | 0.0 |
| Pop Clap, Cybersoda Acoustic Snare, AZ Hi-Hat, Wide Vibrations | within noise | within noise | within noise |

And the CPU side of that trade-off:

| case | clamp off | K=2 | K=0.5 | K=0.125 |
|---|---|---|---|---|
| 1 note, 16 voices, sync 60 | 1.30 | 0.057 | 0.18 | 0.68 |
| 5 notes × 16 voices, sync 60 | 6.52 | 0.27 | 0.89 | 3.41 |

**My recommendation is K=0.5**, which puts the clamp two octaves above where K=2 does. The cost is bounded regardless of sync or modulation, the reported case gets about 7× cheaper, and every bundled patch I rendered stays within about 1 dB. The remaining cost is that an extreme polyphonic chord (5 notes × 16 voices at sync 60) still uses about 0.9 of a core, and for that I think the doctor joke is fair.

An alternative with CPU closer to K=2 and less change in sound: still skip the convolution for steps that land within one clamped interval, but run the correlation recursion for each of them and output their time-weighted mean. That treats the clip-dominated patches honestly. However, cost still grows with sync (with a much smaller constant, since only the RNG and a few multiplies run per skipped step), and some error within the audible band remains. It would need its own prototype to judge.

## Classic: fine

The existing clamp holds. Classic with 16 voices, detune at max, osc octave +3:

| | key 36 | 60 | 84 | 108 | 127 |
|---|---|---|---|---|---|
| sync 0 | 0.007 | 0.008 | 0.014 | 0.034 | 0.040 |
| sync 60 | 0.021 | 0.096 | 0.062 | 0.067 | 0.068 |

Worst case is about 0.1. Nothing to do.

## Wavetable: similar problem, easy fix

Mipmap levels are built all the way down to 2 samples per cycle ([Wavetable.cpp#L430](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/Wavetable.cpp#L430)), but `convolute()` only chooses between levels 0 and 6 ([WavetableOscillator.cpp#L340-L353](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/oscillators/WavetableOscillator.cpp#L340-L353)). A 4096-sample table therefore never goes below 64 steps per cycle, and above a certain pitch the cost grows linearly with pitch. That threshold is roughly 770 Hz for 4096, 1.5 kHz for 2048 (the usual Serum size) and 3 kHz for 1024. Formant doesn't affect cost, since it doesn't change the number of steps per cycle.

The fix is the same selection rule (a step of at least 0.9 samples, a level of at least 2 samples), continued past level 6 for as long as levels exist:

```cpp
while ((mipmap[voice] < max_mipmap_levels - 1) && ((ts >> (mipmap[voice] + 1)) >= 2) &&
       (a * (1 << mipmap[voice]) < 0.5f * wtbias))
    mipmap[voice]++;
```

Saw table, 16 voices, no octave shift, current code → full ladder:

| table | key 96 | key 108 | key 120 | key 127 |
|---|---|---|---|---|
| 1024 | 0.048 → 0.048 | 0.048 → 0.049 | 0.089 → 0.050 | 0.143 → 0.041 |
| 2048 | 0.048 → 0.048 | 0.089 → 0.049 | 0.171 → 0.050 | 0.254 → 0.042 |
| 4096 | 0.089 → 0.048 | 0.170 → 0.049 | 0.335 → 0.050 | 0.497 → 0.042 |

With osc octave +3 on top, a 4096 table reached 1.7 today, which is worse than the S&H case reported here.

Effect on the sound:
- **Below the old cap:** output is identical.
- **Above it:** the main harmonics move by 0.2 dB or less, the top octave loses 1–2 dB, and aliasing residue drops 5–11 dB.
- **Consistency:** today a 4096 table at key 108 is about 1.2 dB brighter at 15–20 kHz than a 1024 table of the same waveform, purely because of the cap. With the fix, large and small tables match.

Only the highest keys change, and the fix makes table sizes consistent with each other, so I think this one is clearly worth doing. It's a separate change from the S&H fix.

## Side findings (separate from this issue)

1. **S&H voices share one random sequence.** `urng = std::bind(distro, storage->rngGen.g)` ([SampleAndHoldOscillator.cpp#L88](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/oscillators/SampleAndHoldOscillator.cpp#L88)) copies the `minstd_rand` by value, so every oscillator starts from a snapshot of the shared generator. Notes started together play the same sequence offset by a few draws, and with retrigger on they're identical. This likely explains why unclamped polyphonic S&H output varies so much between runs.
2. **Absolute unison caps S&H sync at about 40 semitones.** The absolute branch in S&H still floors `t` at 0.1 ([SampleAndHoldOscillator.cpp#L243](https://github.com/surge-synthesizer/surge/blob/6b74d548c1aa3b5e797048930286962135b2df36/src/common/dsp/oscillators/SampleAndHoldOscillator.cpp#L243)). That's the same limit #2115 fixed in Classic by lowering the floor to 0.01.
